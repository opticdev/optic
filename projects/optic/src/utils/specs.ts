import {
  FlatOpenAPIV3,
  FlatOpenAPIV3_1,
  OpenAPIV3,
  defaultEmptySpec,
} from '@useoptic/openapi-utilities';
import { OPTIC_EMPTY_SPEC_KEY, OPTIC_PATH_IGNORE_KEY } from '../constants';
import { JsonSchemaSourcemap } from '@useoptic/openapi-io';
import { logger } from '../logger';

export function getIgnorePaths(
  spec: FlatOpenAPIV3.Document | FlatOpenAPIV3_1.Document
): { path: string; method?: string }[] {
  const ignorePaths: (
    | string
    | {
        method?: string;
        path?: string;
      }
  )[] = spec[OPTIC_PATH_IGNORE_KEY];
  const paths: { path: string; method?: string }[] = [];
  if (!Array.isArray(ignorePaths)) {
    return [];
  } else {
    for (const ignore of ignorePaths) {
      if (typeof ignore === 'string') {
        paths.push({ path: ignore });
      } else if (ignore.method && ignore.path) {
        paths.push({ method: ignore.method, path: ignore.path });
      } else {
        logger.debug(
          `Skipping x-optic-path-ignore ${JSON.stringify(
            ignore
          )} - must be type string or {method: string, path: string}`
        );
      }
    }
  }
  return paths;
}

export function createNewSpecFile(version: string): OpenAPIV3.Document {
  return {
    info: {
      title: 'Untitled service',
      version: '1.0.0',
    },
    openapi: version,
    paths: {},
  };
}

export function createNullSpec(): FlatOpenAPIV3.Document {
  return {
    ...defaultEmptySpec,
    [OPTIC_EMPTY_SPEC_KEY]: true,
  } as FlatOpenAPIV3.Document;
}

export function createNullSpecSourcemap(
  nullSpec: FlatOpenAPIV3.Document
): JsonSchemaSourcemap {
  const emptySpecName = 'empty.json';
  const sourcemap = new JsonSchemaSourcemap(emptySpecName);
  sourcemap.addFileIfMissingFromContents(
    emptySpecName,
    JSON.stringify(nullSpec, null, 2),
    0
  );
  return sourcemap;
}
