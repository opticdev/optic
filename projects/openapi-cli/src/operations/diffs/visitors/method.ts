import { OpenAPIV3 } from '../../../specs';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { Option } from 'ts-results';

export function* visitMethod(
  method: OpenAPIV3.HttpMethods,
  spec: Option<OpenAPIV3.OperationObject>,
  context: { pathPattern: string }
): IterableIterator<OperationDiffResult> {
  if (spec.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedMethod,
      subject: method,
      pathPattern: context.pathPattern,
    };
  }
}
