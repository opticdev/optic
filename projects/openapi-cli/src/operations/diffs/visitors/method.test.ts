import { it, describe, expect } from '@jest/globals';
import { visitMethod } from './method';
import { OpenAPIV3 } from '../../../specs';
import { Some, None } from 'ts-results';

describe('visitMethod', () => {
  it('detects an unspecified path', () => {
    const operation: OpenAPIV3.OperationObject = {
      responses: {},
    };

    const matchingResults = [
      ...visitMethod(OpenAPIV3.HttpMethods.GET, Some(operation), {
        pathPattern: '/orders/{orderId}',
      }),
    ];
    expect(matchingResults).toHaveLength(0);

    const unmatchingResults = [
      ...visitMethod(OpenAPIV3.HttpMethods.GET, None, {
        pathPattern: '/orders/{orderId}',
      }),
    ];
    expect(unmatchingResults).toHaveLength(1);
    expect(unmatchingResults).toMatchSnapshot();
  });
});
