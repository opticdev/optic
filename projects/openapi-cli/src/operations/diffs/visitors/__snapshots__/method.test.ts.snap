// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visitMethod detects an unspecified path 1`] = `
[
  {
    "kind": "UnmatchedMethod",
    "pathPattern": "/orders/{orderId}",
    "subject": "get",
  },
]
`;
